### PR TITLE
[Macros Dialog] Add 2 new parameters: DuplicateIgnoreExtraNote and Du…

### PR DESCRIPTION
**Pull request by mwganson**
…plicateReplaceSpaces

2 more changes, also requested by Mario.

DuplicateIgnoreExtraNote (default = false to keep current behavior by default)

A user may wish to add an extra note to the file name indicating the changes made.

Example: mymacro@005.fix_dialog_font.FCMacro

and have that note ignored in the suggested file name when duplicating this file, so the suggested new name becomes:

mymacro@006.FCMacro instead of mymacro@006.fix_dialog_font.FCMacro

Note: Consider the following list of macro names:

mymacro.FCMacro
mymacro@001.add_cancel_button.FCMacro
mymacro@002.fix_bug_in_object_creation.FCMacro
mymacro@003.resize_dialog.FCMacro

If we duplicate mymacro@002.fix_bug_in_object_creation.FCMacro the suggested name will be mymacro@003.FCMacro and not mymacro@004.FCMacro.  I think this is good because the revisions made in @003 are being tossed out by duplicating the @002 version instead of the @003 version.

Note2: to be recognized as such extra notes must be delimited by decimal (.) after the @NN and before the FCMacro as in:

mymacro@004.some_note.FCMacro

mymacro@004some_note.FCMacro will not be recognized as an extra note and the suggested name is going to be:

mymacro@004some_note@001.FCMacro (which might desired for some users in some cases, similar to branching from a branch)

The other new parameter is DuplicateReplaceSpaces (default = false to keep current behavior by default)

This only happens during the creation of the new duplicate file, not during new macro creation or renaming.  Hence the name, DuplicateReplaceSpaces instead of ReplaceSpaces.  It is for convenience of user to be able to enter spaces instead of harder-to-type underscores.  If true, the spaces entered for the extra note are converted to underscores.  If the user edits the suggested file name by adding a note and the new file name is:

mymacro@004. some extra note.FCMacro

what will be actually used is:

mymacro@004._some_extra_note.FCMacro

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by mwganson** [comment]
Hold committing this.  I want to extend the DuplicateReplaceSpaces to include also new file creation and renaming, rename to ReplaceSpaces, and set default to true.

**Comment by mwganson** [comment]
I have made my changes.  I am finished with this unless someone has additional suggestions.

**Review by chennes — APPROVED** [review]
This compiles, runs, and works as advertised. Since it does change a default behavior I'd like to wait a bit for anyone else who might care to weigh in (either here or in the forums discussion) before merging.

**Comment by yorikvanhavre** [comment]
Looks good to me too! The fact that it modifies an existing behaviour is not very important here IMHO, because it not something heavily used and won't impact what has already been created before.
However, as always when a parameter is not revealed in the UI (for ex. in preferences), it's of uttermost importance to update https://wiki.freecadweb.org/Fine-tuning after the merge

**Comment by mwganson** [comment]
I've added them to the fine tuning page.  I was waiting for this merge to add them all in one go.  This page is a good idea.  I think there are many such parameters that are not on there.

**Comment by yorikvanhavre** [comment]
awesome, thanks!!
yes that might be.. if you see any, go for it! ;)

**Pull request merged**
